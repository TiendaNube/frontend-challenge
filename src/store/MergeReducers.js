import { chain } from "lodash";

const MergeReducers = (reducer) => {
  return (state = {}, action) => {
    return chain(Object.keys(reducer))
      .reduce((init, next) => {
        init[next] = reducer[next](state[next], action);
        return init;
      }, {})
      .value();
  };
};

export default MergeReducers;
