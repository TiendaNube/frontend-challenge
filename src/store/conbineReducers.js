import { chain } from "lodash";

const combineReducers = reducer => {
  return (state = {}, action) => {
    return chain(Object.keys(reducer))
      .reduce((init, next) => {
        init[next] = reducer[next](state[next], action);
        return init;
      }, {})
      .value();
  };
};

export default combineReducers;
