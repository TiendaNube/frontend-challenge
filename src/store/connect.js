import React, { useContext } from "react";
import { useHistory, useLocation, useParams } from "react-router-dom";
import Context from "./context";

const connect = (mapState, mapDispatch) => {
  return WrappedComponent => {
    return props => {
      const { store, dispatch } = useContext(Context);
      return (
        <WrappedComponent
          {...mapState(store)}
          {...mapDispatch(dispatch)}
          {...props}
        />
      );
    };
  };
};

export { connect };
