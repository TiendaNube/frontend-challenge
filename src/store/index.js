import { useContext } from "react";
import Context from "./helpers/Context";

const store = () => {
  const { store, dispatch } = useContext(Context);

  return {
    getState: store,
    dispatch,
  };
};

export default store;
