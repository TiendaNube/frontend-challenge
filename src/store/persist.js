const persist = () => {};

export default persist;
