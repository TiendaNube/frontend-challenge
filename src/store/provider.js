import React, { useState, useReducer, useEffect } from "react";

import PropTypes from "prop-types";
import Context from "./context";
import { STORE } from "./constants/storageKeys";
import { initStoreAction } from "./actions/actionCreators";

const Provider = ({ children, reducer }) => {
  const [store, dispatch] = useReducer(reducer);

  const [state, setState] = useState({ isLoaded: false });

  useEffect(() => {
    dispatch(initStoreAction());
    setState({ isLoaded: true });
  }, []);

  return (
    <Context.Provider displayName="Main Context" value={{ dispatch, store }}>
      {state.isLoaded ? children : false}
    </Context.Provider>
  );
};

Provider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  reducer: PropTypes.func.isRequired,
  persistConfig: PropTypes.shape({
    key: PropTypes.string,
    storage: PropTypes.shape(window.localStorage)
  })
};

Provider.defaultProps = {
  persistConfig: {
    key: STORE,
    localStorage: window.localStorage
  }
};

export default Provider;
