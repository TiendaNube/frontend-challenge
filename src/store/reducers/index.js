import combineReducers from "../conbineReducers";
import productReducer from "./productReducer";
import loginReducers from "./loginReducer";

const reducers = combineReducers({
  productStore: productReducer,
  loginStore: loginReducers
});

export { reducers };
