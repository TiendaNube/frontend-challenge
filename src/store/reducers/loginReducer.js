const initialState = {
  token: null
};

const loginReducers = (state = initialState, action) => {
  switch (action.type) {
    case "LOGIN":
      return {
        ...state,
        token: action.token
      };
    default:
      return state;
  }
};

export default loginReducers;
