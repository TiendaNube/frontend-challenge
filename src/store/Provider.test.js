import React from "react";
import renderer from "react-test-renderer";
import Provider from "store/Provider";
import { reducers } from "store/reducers";

describe("Provider", () => {
  it("render correctly Provider", () => {
    const tree = renderer
      .create(
        <Provider reducer={reducers}>
          <></>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
