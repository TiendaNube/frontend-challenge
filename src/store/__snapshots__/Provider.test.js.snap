// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider render correctly Provider 1`] = `null`;
